pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,621 @@
+					TypeUrl:                obsReq.typeUrl,
+					ResourceNamesSubscribe: obsReq.resourceNames,
+				}
+				log.Debug("Delta send", "req", req)
+				err = delta.Send(req)
+			}
+			if err != nil {
+				log.Error("Stream/Delta send", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+			}
+		case resp, ok := <-c.responseQueue:
+			if !ok {
+				return
+			}
+			log.Debug("Stream receive", "resp", resp)
+			upsertedResources, err := c.handleResponse(resp)
+			if err != nil {
+				log.Error("Failed to handle response", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+				continue
+			}
+			err = c.upsertAndDeleteMissing(upsertedResources)
+			if err != nil {
+				log.Error("Failed to update resources", "error", err)
+				continue
+			}
+			err = c.sendACK(ctx, stream, resp, upsertedResources)
+			if err != nil {
+				log.Error("Failed to send ACK", "error", err)
+				continue
+			}
+		case resp, ok := <-c.deltaResponseQueue:
+			if !ok {
+				return
+			}
+			log.Debug("Delta receive", "resp", resp)
+			upsertedResources, err := c.handleDeltaResponse(resp)
+			if err != nil {
+				log.Error("Failed to handle response", "error", err)
+				errCh <- err
+				err := c.sendDeltaNACK(ctx, delta, resp, err)
+				if err != nil {
+					log.Error("Failed to send NACK", "error", err)
+				}
+				backoff.Wait(ctx)
+				continue
+			}
+			typeUrl := resp.GetTypeUrl()
+			respDelRes := resp.GetRemovedResourceNames()
+			deletedResources := make([]string, 0, len(respDelRes))
+			for _, res := range respDelRes {
+				deletedResources = append(deletedResources, res.GetName())
+			}
+			c.log.Debug("cache TX", "typeUrl", typeUrl, "upserted", upsertedResources, "deleted", deletedResources)
+			ver, updated, _ := c.cache.TX(typeUrl, upsertedResources, deletedResources)
+			c.log.Debug("cache TX", "typeUrl", typeUrl, "ver", ver, "updated", updated)
+			err = c.sendDeltaACK(ctx, delta, resp)
+			if err != nil {
+				log.Error("Failed to send ACK", "error", err)
+				continue
+			}
+		}
+	}
+}
+
+func (c *Client) handleDeltaResponse(resp *discoverypb.DeltaDiscoveryResponse) (map[string]proto.Message, error) {
+	var errs error
+	ret := make(map[string]proto.Message, len(resp.GetResources()))
+	for _, res := range resp.GetResources() {
+		name := res.GetName()
+		msg, _, err := c.parseResource(resp.GetTypeUrl(), res.GetResource())
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+		ret[name] = msg
+	}
+	return ret, errs
+}
+
+// handleResponse decodes the resources based on TypeUrl and updates the cache.
+func (c *Client) handleResponse(resp *discoverypb.DiscoveryResponse) (map[string]map[string]proto.Message, error) {
+	var errs error
+	upsertedResources := make(map[string]map[string]proto.Message, 8)
+	for _, res := range resp.GetResources() {
+		typeUrl := res.GetTypeUrl()
+		msg, name, err := c.parseResource(typeUrl, res)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+		if upsertedResources[typeUrl] == nil {
```

Comment:
minor: you should try to avoid lookup twice in a map. I'm not sure the language allows for the lookups to be optimized away:

```
var typeUrlMap map[string]proto.Message
typeUrlMap, ok := upsertedResources[typeUrl]
if !ok {
  // insert into upsertedResourceMap
  typeUrlMap = ...
}
typeUrlMap[name] = msg
```